### PR TITLE
Fall back to pkgconfig for finding netcdf-cxx4

### PR DESCRIPTION
**Pull request by dschwoerer**
Backport of https://cgit.freebsd.org/ports/tree/science/bout++/files/patch-CMakeLists.txt

@yurivict do you mind checking whether this patch works for you?

### CODE REVIEW

File: cmake/SetupBOUTThirdParty.cmake [hunk 1]
```diff
@@ -167,11 +167,19 @@ if (BOUT_USE_NETCDF)
...
-    find_package(netCDFCxx REQUIRED)
-    set(CONFIG_LDFLAGS "${CONFIG_LDFLAGS} ${netCDF_CXX_LIBRARY} ${netCDF_LIBRARIES}")
+    find_package(netCDFCxx)
+    if (netCDFCxx_FOUND)
+      set(CONFIG_LDFLAGS "${CONFIG_LDFLAGS} ${netCDF_CXX_LIBRARY} ${netCDF_LIBRARIES}")
+      target_link_libraries(bout++ PUBLIC netCDF::netcdf-cxx4)
+    else()
+      find_package(PkgConfig REQUIRED)
+      pkg_check_modules(NETCDF REQUIRED IMPORTED_TARGET netcdf-cxx4)
+      set(CONFIG_LDFLAGS "${CONFIG_LDFLAGS} ${netCDF_CXX_LIBRARY} ${NETCDF_LDFLAGS}")
```
Comment (ZedThree) [inline cmake/SetupBOUTThirdParty.cmake:179]: Where does `netCDF_CXX_LIBRARY` come from in this case?

Comment (dschwoerer) [reply cmake/SetupBOUTThirdParty.cmake:179]: Reading the docs:
https://cmake.org/cmake/help/latest/module/FindPkgConfig.html#command:pkg_check_modules

It should be `NETCDF_LDFLAGS`.

I would assume `netCDF_CXX_LIBRARY` is not set and doesn't do anything ...

File: cmake/SetupBOUTThirdParty.cmake [hunk 2]
```diff
@@ -167,11 +167,19 @@ if (BOUT_USE_NETCDF)
     # Use our own FindnetCDF module which uses nc-config
     find_package(netCDF REQUIRED)
     FetchContent_MakeAvailable(netcdf-cxx4)
+    target_link_libraries(bout++ PUBLIC netCDF::netcdf-cxx4)
   else()
-    find_package(netCDFCxx REQUIRED)
-    set(CONFIG_LDFLAGS "${CONFIG_LDFLAGS} ${netCDF_CXX_LIBRARY} ${netCDF_LIBRARIES}")
+    find_package(netCDFCxx)
+    if (netCDFCxx_FOUND)
+      set(CONFIG_LDFLAGS "${CONFIG_LDFLAGS} ${netCDF_CXX_LIBRARY} ${netCDF_LIBRARIES}")
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply cmake/SetupBOUTThirdParty.cmake:174]: From the docs:
```
#   netCDFCxx_FOUND - true if netCDFCxx was found                               
#   netCDFCxx_VERSION - netCDFCxx version in format Major.Minor.Release         
#   netCDFCxx_INCLUDE_DIRS - Location of the netCDFCxx includes                 
#   netCDFCxx_LIBRARIES - Required libraries   
```
It seems we need `${netCDFCxx_LIBRARIES}` ...
But that is just what was there ... so was it broken before?

File: cmake/SetupBOUTThirdParty.cmake [hunk 3]
```diff
@@ -167,11 +167,19 @@ if (BOUT_USE_NETCDF)
...
+      target_link_libraries(bout++ PUBLIC netCDF::netcdf-cxx4)
+    else()
+      find_package(PkgConfig REQUIRED)
+      pkg_check_modules(NETCDF REQUIRED IMPORTED_TARGET netcdf-cxx4)
+      set(CONFIG_LDFLAGS "${CONFIG_LDFLAGS} ${netCDF_CXX_LIBRARY} ${NETCDF_LDFLAGS}")
+      target_link_libraries(bout++ PUBLIC ${NETCDF_LDFLAGS})
```
Comment (ZedThree) [inline cmake/SetupBOUTThirdParty.cmake:180]: With `IMPORTED_TARGET` in `pkg_check_modules` we can do:

```suggestion
      target_link_libraries(bout++ PUBLIC PkgConfig::netcdf-cxx4)
```

File: cmake/SetupBOUTThirdParty.cmake (hunk 1 above)
Comment (ZedThree) [inline cmake/SetupBOUTThirdParty.cmake:179]:
```suggestion
      set(CONFIG_LDFLAGS "${CONFIG_LDFLAGS} ${NETCDF_LDFLAGS}")
```

File: cmake/SetupBOUTThirdParty.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZedThree) [reply cmake/SetupBOUTThirdParty.cmake:174]: I forget why we're even keeping track of these flags. On my system, only `bout++` and `fmt` end up in `bout-config --libs` and `examples/make-script` works fine. :shrug:

**Review by ZedThree — APPROVED** [review]

File: cmake/SetupBOUTThirdParty.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (dschwoerer) [reply cmake/SetupBOUTThirdParty.cmake:174]: Have you tried building as static lib and than check the make-script?

Should we add a static build to CI? It should be slower :-(
Maybe we should start selecting test cases that we only build on request and once a month?

[the remark this answers is not in the document]
Comment (ZedThree) [reply cmake/SetupBOUTThirdParty.cmake:174]: Oh yes, of course. In a static build, these flags are correct, but for the PkgConfig build, the `LDFLAGS` variable is a CMake-style semicolon-separated list. I'll push a fix

[the remark this answers is not in the document]
Comment (ZedThree) [reply cmake/SetupBOUTThirdParty.cmake:174]: Once we get the CMake-only CI PR in, we should review all the builds

**Review by ZedThree — APPROVED** [review]

**Comment by ZedThree** [comment]
The tests that haven't runner are because they're trying to use the removed 18.04 runners

**Pull request merged**